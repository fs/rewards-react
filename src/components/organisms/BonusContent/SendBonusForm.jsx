import React, { useState } from 'react';
import styled from 'styled-components';
import Button from '../../atoms/Button';
import authService from '../../../services/AuthService';
import bonusService from '../../../services/BonusService';

const Form = styled.form`
  width: 100%;
  background-color: #fff;
  border-radius: 5px;
  margin: 0 0 20px;
  padding: 20px;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.1);
`;

const Textarea = styled.textarea`
  display: block;
  width: 100%;
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #000;
  border: none;
  outline: none;
  padding: 0;
  height: 3.5rem;
  margin: 0 0 20px;
  resize: none;
  box-shadow: none;
`;

const SendBonusForm = (props) => {
  const [bonusText, setBonusText] = useState('');
  const [hasError, setHasError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  const handleChange = (event) => {
    setBonusText(event.target.value);
    setHasError(false);
    setErrorMessage('');
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    const { onSuccess } = props;
    const token = authService.getToken();
    try {
      await bonusService.createBonus(token, bonusText);
      onSuccess();
    } catch (error) {
      const parsedErrorMessage = JSON.parse(error.response.request.response).errors[0].detail;
      setHasError(true);
      setErrorMessage(parsedErrorMessage);
    }
  };

  return (
    <Form onSubmit={handleSubmit} data-test="test-form">
      <Textarea
        name="bonustext"
        onChange={handleChange}
        placeholder="+100 @person add description for #create_awesomness"
        data-testid="test-textarea"
      />
      {hasError && <div className="error-message">{errorMessage}</div>}
      <Button text="Give" data-testid="test-button" />
    </Form>
  );
};

export default SendBonusForm;
