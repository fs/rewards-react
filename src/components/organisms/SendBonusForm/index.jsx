import React, { Component } from 'react';
import styled from 'styled-components';
import Button from '../../atoms/Button';
import bonusService from '../../../services/BonusService';
import authService from '../../../services/AuthService';

const Form = styled.form`
  width: 100%;
  background-color: #fff;
  border-radius: 5px;
  padding: 20px;
`;

const Textarea = styled.textarea`
  display: block;
  width: 100%;
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #000;
  border: none;
  outline: none;
  padding: 0;
  height: 3.5rem;
  margin: 0 0 30px 0;
  resize: none;
  box-shadow: none;
`;

class SendBonusForm extends Component {
  state = {
    bonusText: '',
    hasError: false,
    errorMessage: '',
  };

  handleChange = event => {
    this.setState({ bonusText: event.target.value });
  };

  handleSubmit = async event => {
    event.preventDefault();
    const { bonusText } = this.state;
    const token = authService.getToken();
    try {
      await bonusService.createBonus(token, bonusText);
    } catch (error) {
      const errorMessage = JSON.parse(error.response.request.response).errors[0].detail;
      this.setState({
        hasError: true,
        errorMessage,
      });
    }
  };

  render() {
    return (
      <Form onSubmit={this.handleSubmit}>
        <Textarea
          name="bonustext"
          onChange={this.handleChange}
          placeholder="+100 @person add description for #create_awesomness"
        />
        {this.state.hasError && <div className="error-message">{this.state.errorMessage}</div>}
        <Button text="Give" />
      </Form>
    );
  }
}

export default SendBonusForm;
