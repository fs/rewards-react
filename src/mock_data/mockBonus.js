export default {
  comments: [],
  'created-at': '2019-08-28T12:29:11.030Z',
  id: '558',
  points: 1,
  receivers: [
    {
      'allowance-balance': 490,
      'bonus-balance': 97,
      email: 'albert.fazullin@flatstack.com',
      'full-name': 'Albert Fazullin',
      'profile-image-avatar-url':
        'http://doypx5fd26upp.cloudfront.net/assets/default-user-profile_image-b50dbea6a3e28a1ae62538d4b0115b18558fac3ff7dbbee7da800b5216f97070.png',
      username: 'albert.fazullin',
    },
  ],
  sender: 'Nadezhda Kharchuk',
  text: [
    {
      text: '+1',
      type: 'points',
    },
    {
      text: '@albert.fazullin',
      type: 'users',
    },
    {
      text: '#create-awesomeness',
      type: 'tags',
    },
  ],
  'total-points': 1,
};
