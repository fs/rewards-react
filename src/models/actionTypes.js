export const UPDATE_USER_LOADING = 'UPDATE_USER_LOADING';
export const UPDATE_USER_ERROR = 'UPDATE_USER_ERROR';
export const UPDATE_USER_SUCCESS = 'UPDATE_USER_SUCCESS';
export const UPDATE_BONUS_LIST_SUCCESS = 'UPDATE_BONUS_LIST_SUCCESS';
export const UPDATE_BONUS_LIST_ERROR = 'UPDATE_BONUS_LIST_ERROR';
export const UPDATE_BONUS_LIST_LOADING = 'UPDATE_BONUS_LIST_LOADING';
