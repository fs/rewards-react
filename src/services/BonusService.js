import api from './ApiService';

const apiUrl = '/user/bonuses';

export default class BonusService {
  static async createBonus(token, text) {
    const config = {
      headers: { Authorization: `Bearer ${token}` },
    };

    const bodyParameters = {
      data: {
        type: 'bonus-texts',
        attributes: {
          text,
        },
      },
    };

    const response = await api.post(apiUrl, bodyParameters, config);
    return response;
  }

  static async updateBonusesList(token) {
    const config = {
      headers: { Authorization: `Bearer ${token}` },
    };

    const response = await api.get(apiUrl, config);
    return response;
  }
}
