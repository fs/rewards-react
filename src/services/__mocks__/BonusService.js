export default class BonusService {
  static async createBonus() {
    console.log('createBonus');
    return {};
  }

  static async fetchBonusesList() {
    console.log('fetchBonusesList');
    return {};
  }
}
