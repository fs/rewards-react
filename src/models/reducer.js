import * as types from './actionTypes';
import bonusParser from '../utils/bonusParser';

const reducer = (state, action) => {
  switch (action.type) {
    case types.UPDATE_USER_SUCCESS:
      return {
        ...state,
        user: action.payload,
        isUserLoading: false,
        hasUserError: false,
      };
    case types.UPDATE_USER_LOADING:
      return {
        ...state,
        isUserLoading: true,
        hasUserError: false,
      };
    case types.UPDATE_USER_ERROR:
      return {
        ...state,
        isUserLoading: false,
        hasUserError: true,
      };
    case types.UPDATE_BONUS_LIST_SUCCESS:
      return {
        ...state,
        bonusList: action.payload,
        isBonusListLoading: false,
        hasBonusListError: false,
      };
    case types.UPDATE_BONUS_LIST_ERROR:
      return {
        ...state,
        hasBonusListError: true,
        isBonusListLoading: false,
      };
    case types.UPDATE_BONUS_LIST_LOADING:
      return {
        ...state,
        isBonusListLoading: true,
        hasBonusListError: false,
      };
    case types.UPDATE_BONUS_LIST_AFTER_ADD_COMMENT_SUCCESS: {
      const newBonus = bonusParser(action.payload);
      const { bonusList } = state;

      const updatedBonusList = bonusList.map(item => {
        if (item.id === newBonus[0].id) {
          return newBonus[0];
        }
        return item;
      });

      console.log('action', updatedBonusList);

      return {
        ...state,
        bonusList: updatedBonusList,
      };
    }
    default:
      return state;
  }
};

export default reducer;
