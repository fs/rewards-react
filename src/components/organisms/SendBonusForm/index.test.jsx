import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';

import BonusService from '../../../services/BonusService';
import Context from '../../context/Context';
import SendBonusForm from '.';
import * as types from '../../../models/actionTypes';

import mockBonus from '../../../mock_data/mockBonus';

jest.mock('../../../services/BonusService');

describe('SendBonusForm', () => {
  const dispatch = jest.fn();

  beforeEach(() => {
    jest.resetModules();
    jest.clearAllMocks();
  });

  test('should call BonusService.createBonus on submit', async () => {
    // Arrange
    const expectedBonusCount = '+1';
    const expectedReceiver = '@albert.fazullin';
    const expectedTag = '#create-awesomeness';
    const expectedBonusText = `${expectedBonusCount} ${expectedReceiver} ${expectedTag}`;

    const expectedNewBonusAction = {
      type: types.UPDATE_BONUS_LIST_AFTER_ADD_BONUS_SUCCESS,
      payload: mockBonus,
    };

    const expectedAllowanceBalanceAction = {
      type: types.UPDATE_ALLOWANCE_BALANCE,
      payload: mockBonus.sender['allowance-balance'],
    };

    const mockCreateBonus = jest.fn(() => Promise.resolve(mockBonus));

    BonusService.createBonus.mockImplementation(mockCreateBonus);

    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <SendBonusForm />,
      </Context.Provider>,
    );
    const textArea = getByTestId('test-textarea');
    fireEvent.change(textArea, { target: { value: expectedBonusText } });
    const form = getByTestId('test-bonus-form');

    // Act
    fireEvent.submit(form);

    // Assert
    await wait(() => {
      expect(BonusService.createBonus).toHaveBeenCalledWith(expectedBonusText);
    });

    expect(dispatch).toHaveBeenNthCalledWith(1, expectedNewBonusAction);
    expect(dispatch).toHaveBeenNthCalledWith(2, expectedAllowanceBalanceAction);
  });

  test('should call bonusService.createBonus on submit and show error message when not enough points', async () => {
    // Arrange
    const expectedBonusCount = '+10000000000';
    const expectedReceiver = '@albert.fazullin';
    const expectedTag = '#create-awesomeness';
    const expectedBonusText = `${expectedBonusCount} ${expectedReceiver} ${expectedTag}`;

    const expectedErrorMessage = "You can't give more points than 500.";
    const expectedResponse = `{"errors":[{"source":{"pointer":"/data/attributes/base"},"detail":"${expectedErrorMessage}"}]}`;
    const expectedError = { response: { request: { response: expectedResponse } } };

    const mockCreateBonus = jest.fn(
      () =>
        new Promise((resolve, reject) => {
          reject(expectedError);
        }),
    );
    BonusService.createBonus.mockImplementation(mockCreateBonus);

    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <SendBonusForm />,
      </Context.Provider>,
    );
    const textArea = getByTestId('test-textarea');

    fireEvent.change(textArea, { target: { value: expectedBonusText } });

    const form = getByTestId('test-bonus-form');
    const errorContainer = getByTestId('test-error-container');

    // Act
    fireEvent.submit(form);

    // Assert
    await wait(() => {
      expect(BonusService.createBonus).toHaveBeenCalledWith(expectedBonusText);
      expect(errorContainer).toHaveTextContent(expectedErrorMessage);
    });
  });
});

describe('HelperIcon', () => {
  const dispatch = jest.fn();

  test('Should change the className on user change in textarea value', () => {
    // Arrange
    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <SendBonusForm />,
      </Context.Provider>,
    );
    const textArea = getByTestId('test-textarea');
    const expectedBonusText = '#create-awesomeness';

    // Act
    fireEvent.change(textArea, { target: { value: expectedBonusText } });

    const bonusForm = getByTestId('test-bonus-form');
    // Assert
    expect(bonusForm).toMatchSnapshot();
  });
});
