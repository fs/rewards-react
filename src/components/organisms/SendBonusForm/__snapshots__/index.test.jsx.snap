// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendBonusForm HelperIcon Should change the className on user change in textarea value 1`] = `
.c2 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  overflow: hidden;
  background-color: #d8d8d8;
  margin-right: 10px;
  cursor: pointer;
  -webkit-transition: box-shadow 0.08s linear;
  transition: box-shadow 0.08s linear;
  outline: none;
  border: 0;
}

.c2:hover,
.c2.active {
  background-color: rgba(0,85,112,0.7);
}

.c2:focus {
  box-shadow: 0 1px 3px 0 rgba(60,64,67,0.302),0 4px 8px 3px rgba(60,64,67,0.149);
}

.c2:last-child {
  margin-right: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 20px;
}

.c3 {
  background-color: #63bc36;
  color: #fff;
  border-radius: 100px;
  height: 40px;
  line-height: 40px;
  padding: 0 20px;
  font-size: 16px;
  font-weight: bold;
  border: none;
}

.c3:hover,
.c3:focus {
  outline: none;
  background-color: #4f962b;
}

.c0 {
  width: 100%;
  background-color: #fff;
  border-radius: 5px;
  margin: 0 0 20px;
  padding: 20px;
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.1);
}

<form
  class="c0"
  data-testid="test-bonus-form"
>
  <div
    data-testid="test-textarea-wrapper"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
        type="button"
      >
        <img
          alt="Points"
          src="[object Object]"
        />
      </button>
      <button
        class="c2"
        type="button"
      >
        <img
          alt="User"
          src="[object Object]"
        />
      </button>
      <button
        class="active c2"
        type="button"
      >
        <img
          alt="Hashtag"
          src="[object Object]"
        />
      </button>
    </div>
    <div
      class="rta  "
    >
      <textarea
        class="rta__textarea autocomplete-textarea"
        data-testid="test-textarea"
        name="bonus-text"
        placeholder="+100 @person add description for #create_awesomness"
      >
        #create-awesomeness
      </textarea>
    </div>
  </div>
  <div
    data-testid="test-error-container"
  />
  <button
    class="c3"
    data-testid="test-button"
  >
    Give
  </button>
</form>
`;
